### ISSUE TITLE
How to forward address of the worker node using port-forwarding                                                                                                                                                                                                                                                                                                                                                            

### ISSUE THREAD

**Issue by uchiha-pain** [canonical/microk8s#3292]
### My setup:
I have a microk8s cluster with 8 nodes. The formation of the cluster consist of 2 master nodes (control plane), and 6 worker nodes.
```
NAME            STATUS   ROLES    AGE     VERSION                    LABELS
controlplan1    Ready    <none>   2d2h    v1.24.0-2+59bbb3530b6769   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=controlplan1,kubernetes.io/os=linux,microk8s.io/cluster=true,node.kubernetes.io/microk8s-controlplane=microk8s-controlplane
medusa          Ready    <none>   2d1h    v1.24.0-2+59bbb3530b6769   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=medusa,kubernetes.io/os=linux,microk8s.io/cluster=true,node.kubernetes.io/microk8s-worker=microk8s-worker
circe           Ready    <none>   2d1h    v1.24.0-2+59bbb3530b6769   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=circe,kubernetes.io/os=linux,microk8s.io/cluster=true,node.kubernetes.io/microk8s-worker=microk8s-worker
lachesis        Ready    <none>   2d1h    v1.24.0-2+59bbb3530b6769   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=lachesis,kubernetes.io/os=linux,microk8s.io/cluster=true,node.kubernetes.io/microk8s-worker=microk8s-worker
atropos         Ready    <none>   2d1h    v1.24.0-2+59bbb3530b6769   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=atropos,kubernetes.io/os=linux,microk8s.io/cluster=true,node.kubernetes.io/microk8s-worker=microk8s-worker
clotho          Ready    <none>   2d1h    v1.24.0-2+59bbb3530b6769   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=clotho,kubernetes.io/os=linux,microk8s.io/cluster=true,node.kubernetes.io/microk8s-worker=microk8s-worker
calypso         Ready    <none>   2d1h    v1.24.0-2+59bbb3530b6769   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=calypso,kubernetes.io/os=linux,microk8s.io/cluster=true,node.kubernetes.io/microk8s-worker=microk8s-worker
controlplane1   Ready    <none>   2d21h   v1.24.0-2+59bbb3530b6769   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=controlplane1,kubernetes.io/os=linux,microk8s.io/cluster=true,node.kubernetes.io/microk8s-controlplane=microk8s-controlplane
```
The further break down of the cluster is as follows
controlplane1(192.168.1.81) is the main master node and is comprised of 3 worker nodes: atropos(192.168.1.83), clotho(192.168.1.84), and lachesis(192.168.1.85).
controlplan1(192.168.1.82) is the 2nd master node and is comprised of other 3 nodes: circe(192.168.1.86), calypso(192.168.1.87), and medusa(192.168.1.88).

### What I want:
Currently I have Cassandra database pods deployed on these worker nodes. I want to implement such strategy that if my both control-planes go down then I still want to be able to connect to my Cassandra pods because they are still up and running. I read somewhere that if control-plane goes down then the worker nodes should still perform their job that was defined by the control-plan earlier.

### What I am doing:
I am forwarding the address of the both control-planes with port forwarding command.
`microk8s kubectl -n control-plane-1 port-forward --address 192.168.1.81 pod/demo-dc1-default-sts-0 44441:9042`
`microk8s kubectl -n control-plane-2 port-forward --address 192.168.1.82 pod/demo-dc1-default-sts-0 44442:9042`

These command are working fine but when my control-planes go down then I lose access to the whole datacenter. Then I thought what if I forward the address of the worker node instead of the control-plane. In this way, if my control-plane goes down then I will still have the IP for external connection. 

### What I am getting:
When I run the port forward command (`microk8s kubectl -n control-plane-2 port-forward --address 192.168.1.83 pod/demo-dc2-default-sts-0 44441:9042`) from the worker node then I get the following error:
`This MicroK8s deployment is acting as a node in a cluster. Please use the microk8s kubectl on the master.`
and
When I run the above command from the control-plane then I get the following error
```
Unable to listen on port 44441: Listeners failed to create with the following errors: [unable to create listener: Error listen tcp4 192.168.1.83:44441: bind: cannot assign requested address]
error: unable to listen on any of the requested ports: [{44441 9042}]
```
And if use sudo before microk8s `sudo microk8s kubectl -n control-plane-1 port-forward --address 192.168.1.85 pod/demo-dc1-default-sts-0 44441:9042` then I get the following stopped job
```
jobs -l
[1]+ 38319 Stopped (tty output)    sudo microk8s kubectl -n control-plane-1 port-forward --address 192.168.1.85 pod/demo-dc1-default-sts-0 44441:9042
```
Is there anything that I am not doing correctly or there is an other way to do this?
Thanks
Uchiha Pain!

**Comment by neoaggelos** [comment]
Hi @uchiha-pain, this looks more like a question than an issue.

**TL;DR**: You should be using `Service` definitions to access Cassandra from outside the cluster. See [Service | Kubernetes](https://kubernetes.io/docs/concepts/services-networking/service/) for details.

**Longer version**:

I will try and address some of your points inline

> I read somewhere that if control-plane goes down then the worker nodes should still perform their job that was defined by the control-plan earlier.

This is correct, the Kubernetes control plane is running independently from the workloads. Even if the control plane goes down, this should not affect the workloads.

> `microk8s kubectl -n control-plane-1 port-forward --address 192.168.1.81 pod/demo-dc1-default-sts-0 44441:9042`

Port forwarding is mostly meant for local development and debugging purposes, not for accessing production workloads. For this, you should instead be using `Service` definitions. Reading your post, the easiest way to get started is to use a [`NodePort` service](https://kubernetes.io/docs/concepts/services-networking/service/#type-nodeport). This will expose a specific port from your workloads on a specific high port on all your cluster nodes (both control plane, and worker nodes).

This works great if you already have a "load balancer" of sorts outside of the cluster. Alternatively, a more complex setup would be to configure [MetalLB](https://microk8s.io/docs/addon-metallb) and use a [`LoadBalancer` service](https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer).

I would suggest starting simple and seeing what works for you. Feel free to keep the discussion going here, or in the [#microk8s channel on Slack](https://kubernetes.slack.com/archives/CAUNWQ85V). Welcome to Kubernetes!
